src/Domain/Block.hpp

```diff
@@ -0,0 +1,89 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class template Block.
+
+#pragma once
+
+#include <iosfwd>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include "Domain/BlockNeighbor.hpp"
+#include "Domain/Direction.hpp"
+#include "Domain/EmbeddingMaps/EmbeddingMap.hpp"
+
+/// \ingroup ComputationalDomain
+/// A Block<VolumeDim> is a region of a VolumeDim-dimensional computational
+/// domain that defines the root node of a tree which is used to construct the
+/// Elements that cover the computational domain.
+///
+/// Each codiminsion 1 boundary of a Block<VolumeDim> is either an external
+/// boundary or identical to a boundary of one other Block.
+///
+/// A Block has logical coordinates that go from -1 to +1 in each
+/// dimension.  The global coordinates are obtained from the logical
+/// coordinates from the EmbeddingMap:  EmbeddingMap::operator() takes
+/// logical coordinates and returns global coordinates.
+///
+/// \tparam VolumeDim the dimension of the Block (i.e. the volume dimension).
+template <size_t VolumeDim>
+class Block {
+  static constexpr int private_asserts() {
+    static_assert(std::is_nothrow_move_constructible<Block>::value and
+                      std::is_nothrow_move_assignable<Block>::value,
+                  "Missing move semantics");
+    return 0;
+  }
+
+ public:
+  /// Construct a Block.
+  ///
+  /// \param embedding_map the EmbeddingMap.
+  /// \param my_id a unique ID.
+  /// \param neighbors info about the Blocks abutting this Block.
+  /// \param external_boundaries the directions in which there are external
+  ///           boundaries
+  Block(const EmbeddingMap<VolumeDim, VolumeDim>& embedding_map, size_t id,
+        std::unordered_map<Direction<VolumeDim>, BlockNeighbor<VolumeDim>>&&
```

Comment:
If we want to force people to `std::move` their `unordered_map` into the constructor because we cannot see any case where we should not be taking ownership of the data then `&&`. If there would be a reason to not move the `unordered_map` in, but instead pass in a copy and we think this should be the default thing to do, then take by value. I don't know enough about the code that calls the constructor to tell you what is the right answer. Is that enough to help you make the decision?